sale_commission_product_criteria/models/commission.py

```diff
@@ -0,0 +1,392 @@
+    _inherit = "sale.commission.line.mixin"  # OCA sale_commission
+
+    applied_commission_id = fields.Many2one("sale.commission", readonly=True)
+    commission_id = fields.Many2one(
+        comodel_name="sale.commission",
+        ondelete="restrict",
+        required=False,
+        compute="_compute_commission_id",
+        store=True,
+        readonly=False,
+        copy=True,
+    )
+
+    def _get_commission_items(self, commission, product):
+        # Method replaced
+        categ_ids = {}
+        categ = product.categ_id
+        while categ:
+            categ_ids[categ.id] = True
+            categ = categ.parent_id
+        categ_ids = list(categ_ids)
+        # Select all suitable items. Order by best match
+        # (priority is: all/cat/subcat/product/variant).
+        self.env.cr.execute(
+            """
+            SELECT
+                item.id
+            FROM
+                commission_item AS item
+            LEFT JOIN product_category AS categ ON item.categ_id = categ.id
+            WHERE
+                (item.product_tmpl_id IS NULL OR item.product_tmpl_id = any(%s))
+                AND (item.product_id IS NULL OR item.product_id = any(%s))
+                AND (item.categ_id IS NULL OR item.categ_id = any(%s))
+                AND (item.commission_id = %s)
+            ORDER BY
+                item.applied_on, categ.complete_name desc, item.id desc
+            """,
+            (
+                product.product_tmpl_id.ids,
+                product.ids,
+                categ_ids,
+                commission._origin.id,  # Added this
+            ),
+        )
+        item_ids = [x[0] for x in self.env.cr.fetchall()]
+        return item_ids
+
+    def _get_single_commission_amount(self, commission, subtotal, product, quantity):
+        self.ensure_one()
+        item_ids = self._get_commission_items(commission, product)
+        if not item_ids:
+            return 0.0
+        commission_item = self.env["commission.item"].browse(item_ids[0])
+        if commission.amount_base_type == "net_amount":
+            # If subtotal (sale_price * quantity) is less than
+            # standard_price * quantity, it means that we are selling at
+            # lower price than we bought, so set amount_base to 0
+            subtotal = max([0, subtotal - product.standard_price * quantity])
+        self.applied_commission_item_id = commission_item
+        # if self.agent_id.use_multi_type_commissions:
+        self.applied_commission_id = commission_item.commission_id
+        if commission_item.commission_type == "fixed":
+            return commission_item.fixed_amount
+        elif commission_item.commission_type == "percentage":
+            return subtotal * (commission_item.percent_amount / 100.0)
+
+
+class SaleCommission(models.Model):
+    _inherit = "sale.commission"
+
+    commission_type = fields.Selection(
+        selection_add=[("rules", "Rules")], ondelete={"rules": "cascade"}
```

Comment:
Put a better name and don't delete on cascade (it will provoke problems on existing invoice lines):

```suggestion
        selection_add=[("product", "Product criteria")], ondelete={"rules": "set default"}
```